File: C++/023. Merge k Sorted Lists.cpp
#include <iostream>
#include <vector>

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* mergeKLists(std::vector<ListNode*>& lists) {
        if (lists.size() == 0)
            return nullptr;
        while (int(lists.size()) > 1) {
            lists.push_back(merge2Lists(lists[0], lists[1]));
            lists.erase(lists.begin(), lists.begin() + 2);
        }
        return lists[0];
    }

    ListNode* merge2Lists(ListNode* l1, ListNode* l2) {
        ListNode *res = new ListNode(0), *p = res;
        while (l1 != nullptr && l2 != nullptr) {
            if (l1->val < l2->val) {
                p->next = l1;
                l1 = l1->next;
            } else {
                p->next = l2;
                l2 = l2->next;
            }
            p = p->next;
        }
        if (l1 == nullptr)
            p->next = l2;
        else
            p->next = l1;
        return res->next;
    }
};

File: Go/023. Merge k Sorted Lists.go
package main

type ListNode struct {
	Val int
	Next *ListNode
}

func mergeKLists(lists []*ListNode) *ListNode {
	if len(lists) == 0 {
		return nil
	}
	for len(lists) > 1 {
		lists = append(lists, merge2Lists(lists[0], lists[1]))
		lists = lists[2 :]
	}
	return lists[0]
}

func merge2Lists(l1 *ListNode, l2 *ListNode) *ListNode {
	res := new(ListNode)
	p := res
	for l1 != nil && l2 != nil {
		if l1.Val < l2.Val {
			p.Next = l1
			l1 = l1.Next
		} else {
			p.Next = l2
			l2 = l2.Next
		}
		p = p.Next
	}
	if l1 == nil {
		p.Next = l2
	} else {
		p.Next = l1
	}
	return res.Next
}